### PR TITLE
Replace exceptions related to event subscriptions with either

### PR DESCRIPTION
**Pull request by korthout**
## Description

This replaces the subscribe to message and timer events logic with functional style using the Either type.

To achieve that, this PR adds to the previously added `Either.collect()` a `Either.collectFoldingLeft` to collect multiple calculations (evaluations of the expressions) that can fail to return either a failure or subscribe to all events using the evaluated expressions.

It's likely best to review this on a commit by commit basis. However, please keep in mind that not all changes made in the first commits are kept in the final result.

## Related issues

relates #4323

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by korthout** [comment]
@saig0 Please have a look, and let me know if you see ways to improve this further. In addition, if you want to me extract the Either.peek to a separate pull request, please let me know.

**Review by saig0 — CHANGES_REQUESTED** [review]
I like the new style :+1: Thanks @korthout :cake: 

I just have some minor comments. Please have a look.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java [hunk 1]
```diff
@@ -90,45 +90,54 @@ public void unsubscribeFromEvents(
...
-                }));
-
-    // we can only deal with the first failure, so it's enough to return that
-    return failureOrEvalResult
+    return supplier.getEvents().stream()
+        .filter(event -> event.isTimer() || event.isMessage())
+        .map(event -> evalExpressions(event, context))
+        .collect(Either.collector())
+        .peek(results -> subscribeToMessageEvents(context, sideEffects, results))
+        .peek(results -> subscribeToTimerEvents(context, sideEffects, commandWriter, results))
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java:98]: That's an interesting point. I guess the confusion during yesterday's show and tell was partially due to the peek and could've been avoided this way. 

I do feel it takes the power of chaining away, which feels very useful in the java world where side-effects are everywhere. Another idea is to change the Either.ifRight signature to what peek actually is. This allows something like the following.

```suggestion
        .ifRight(results -> subscribeToMessageEvents(context, sideEffects, results))
        .ifRight(results -> subscribeToTimerEvents(context, sideEffects, commandWriter, results))
```

 To be fair, probably it's best not allow this chaining altogether and make side effects more visible by having to either place it in the end of a chain, or separate it like in your example. I'll remove the `peek` for now

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java [hunk 2]
```diff
@@ -229,7 +237,17 @@ public ExpressionProcessor(
...
+            result ->
+                typeCheck(result, expectedTypes, scopeKey)
+                    .mapLeft(
+                        failure ->
+                            new Failure(
+                                String.format(
+                                    "Failed to extract the correlation key for '%s': The value must be either a string or a number, but was %s.",
+                                    expression.getExpression(), result.getType()),
+                                ErrorType.EXTRACT_VALUE_ERROR,
+                                scopeKey)))
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java:250]: @saig0 Note that this is the only place where we deviate from the standard typecheck failure message. If we align the messages this entire mapLeft is unnecessary. WDYT?

**Comment by korthout** [comment]
@saig0 I've played around a bit more after your comments. I've added a collectorFoldingLeft Collector in an effort to reduce the required logic, removed the peek method to highlight the side effects and reduced the return value mapping logic to the bare minimum.

Please have another look.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java:250]: Yes, I see your point. I would like to be consistent too.

The advantage of this message is that it says what is wrong. For example, if the incident is created for a task with multiple boundary events. 

We could change the standard type check and add a context to it (e.g. `... while evaluating the correlation key`).

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/common/CatchEventBehavior.java]: > probably it's best not allow this chaining altogether and make side effects more visible by having to either place it in the end of a chain

I agree. 

Allow to chain `ifRight()` would also differ from the Java API of `Optional#ifPresent()` and the semantics of terminal stream operations.

**Review by saig0 — APPROVED** [review]
Looks good :+1:

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java:250]: Interesting. I'll just open an issue for that idea. I'll extract the special code in a separate method for now

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
bors cancel

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
bors retry

**Pull request merged**